sql: valueNode created for virtual tables not properly released upon error

```
I161010 14:33:20.777685 1 util/log/clog.go:985  [config] file created at: 2016/10/10 14:33:20
I161010 14:33:20.777685 1 util/log/clog.go:985  [config] running on machine: Alexs-MacBook-Air
I161010 14:33:20.777685 1 util/log/clog.go:985  [config] binary: CockroachDB beta-20161006-796-gdc05d14 (darwin amd64, built 2016/10/10 18:27:14, go1.7.1)
I161010 14:33:20.777685 1 util/log/clog.go:985  [config] arguments: [./cockroach start]
```

- What did you do?

Started a server, then attempted to run the Django tests. They errored out with:

`unknown function: pg_catalog.pg_table_is_visible`

- What did you expect to see?

The server to still be running.

- What did you see instead?

A `panic`:

```
panic: unexpected leftover memory: 16601 bytes [recovered]
	panic: unexpected leftover memory: 16601 bytes

goroutine 168 [running]:
panic(0x505a500, 0xc42060a9d0)
	/usr/local/Cellar/go/1.7.1/libexec/src/runtime/panic.go:500 +0x1a1
github.com/cockroachdb/cockroach/util/stop.(*Stopper).Recover(0xc42008cd80)
	/private/tmp/x/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:180 +0x6e
panic(0x505a500, 0xc42060a9d0)
	/usr/local/Cellar/go/1.7.1/libexec/src/runtime/panic.go:458 +0x243
github.com/cockroachdb/cockroach/sql/mon.(*MemoryUsageMonitor).StopMonitor(0xc4219fc7f0, 0x7290520, 0xc420c3c6c0)
	/private/tmp/x/src/github.com/cockroachdb/cockroach/sql/mon/mem_usage.go:130 +0x2a2
github.com/cockroachdb/cockroach/sql.(*Session).Finish(0xc4219fc000)
	/private/tmp/x/src/github.com/cockroachdb/cockroach/sql/session.go:185 +0xac
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).finish(0xc420118900)
	/private/tmp/x/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:151 +0x7f
github.com/cockroachdb/cockroach/sql/pgwire.(*Server).ServeConn(0xc420239680, 0x59c41c0, 0xc4205fee70, 0x59a8680, 0xc420012140)
	/private/tmp/x/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:242 +0x652
github.com/cockroachdb/cockroach/server.(*Server).Start.func8.1(0x59c41c0, 0xc4205fee70)
	/private/tmp/x/src/github.com/cockroachdb/cockroach/server/server.go:454 +0x4e
github.com/cockroachdb/cockroach/util/netutil.(*Server).ServeWith.func1(0xc42008cd80, 0xc42012a298, 0x59c41c0, 0xc4205fee70, 0xc420128080)
	/private/tmp/x/src/github.com/cockroachdb/cockroach/util/netutil/net.go:136 +0x95
created by github.com/cockroachdb/cockroach/util/netutil.(*Server).ServeWith
	/private/tmp/x/src/github.com/cockroachdb/cockroach/util/netutil/net.go:138 +0x277
```

--- a/sql/data_source.go
+++ b/sql/data_source.go
@@ -199,18 +199,19 @@ func (p *planner) getVirtualDataSource(tn *parser.TableName) (planDataSource, bo
 		return planDataSource{}, false, err
 	}
 	if virtual.desc != nil {
-		v, err := virtual.getValuesNode(p)
-		if err != nil {
-			return planDataSource{}, false, err
-		}
-
+		columns, constructor := virtual.getPlanInfo()
 		sourceName := parser.TableName{
 			TableName:    parser.Name(virtual.desc.Name),
 			DatabaseName: tn.DatabaseName,
 		}
 		return planDataSource{
-			info: newSourceInfoForSingleTable(sourceName, v.Columns()),
-			plan: v,
+			info: newSourceInfoForSingleTable(sourceName, columns),
+			plan: &delayedValuesNode{
+				p:           p,
+				name:        sourceName.String(),
+				columns:     columns,
+				constructor: constructor,
+			},
 		}, true, nil
 	}
 	return planDataSource{}, false, nil

--- a/sql/plan.go
+++ b/sql/plan.go
@@ -212,6 +212,7 @@ var _ planNode = &dropViewNode{}
 var _ planNode = &alterTableNode{}
 var _ planNode = &joinNode{}
 var _ planNode = &distSQLNode{}
+var _ planNode = &delayedValuesNode{}
 
 // makePlan implements the Planner interface.
 func (p *planner) makePlan(stmt parser.Statement, autoCommit bool) (planNode, error) {

--- a/sql/testdata/pg_catalog
+++ b/sql/testdata/pg_catalog
@@ -24,6 +24,11 @@ CREATE DATABASE pg_catalog
 statement ok
 DROP DATABASE other_db
 
+# the following query checks that the planDataSource instantiated from
+# a virtual table in the FROM clause is properly deallocated even when
+# query preparation causes an error. #9853
+query error unknown function
+SELECT * FROM pg_catalog.pg_class c WHERE nonexistent_function()
 
 # Verify pg_catalog can be used like a normal database.
 

--- a/sql/values.go
+++ b/sql/values.go
@@ -318,3 +318,53 @@ func (n *valuesNode) ExplainTypes(regTypes func(string, string)) {
 }
 
 func (*valuesNode) SetLimitHint(_ int64, _ bool) {}
+
+// delayedValuesNode wraps a valuesNode in cases where the valuesNode
+// must be populated during query execution (as opposed to planNode
+// construction) for resource tracking purposes.
+type delayedValuesNode struct {
+	p           *planner
+	name        string
+	columns     ResultColumns
+	constructor valuesNodeConstructor
+	plan        *valuesNode
+}
+
+func (d *delayedValuesNode) SetLimitHint(_ int64, _ bool) {}
+func (d *delayedValuesNode) expandPlan() error {
+	v, err := d.constructor(d.p)
+	if err != nil {
+		return err
+	}
+	if err := v.expandPlan(); err != nil {
+		v.Close()
+		return err
+	}
+	d.plan = v
+	return nil
+}
+
+func (d *delayedValuesNode) Close() {
+	if d.plan != nil {
+		d.plan.Close()
+		d.plan = nil
+	}
+}
+
+func (d *delayedValuesNode) ExplainPlan(
+	verbose bool,
+) (name, description string, children []planNode) {
+	if d.plan != nil {
+		children = []planNode{d.plan}
+	}
+	return "virtual table", d.name, children
+}
+
+func (d *delayedValuesNode) ExplainTypes(rt func(string, string)) {}
+func (d *delayedValuesNode) Columns() ResultColumns               { return d.columns }
+func (d *delayedValuesNode) Ordering() orderingInfo               { return orderingInfo{} }
+func (d *delayedValuesNode) MarkDebug(_ explainMode)              {}
+func (d *delayedValuesNode) Start() error                         { return d.plan.Start() }
+func (d *delayedValuesNode) Next() (bool, error)                  { return d.plan.Next() }
+func (d *delayedValuesNode) Values() parser.DTuple                { return d.plan.Values() }
+func (d *delayedValuesNode) DebugValues() debugValues             { return d.plan.DebugValues() }

--- a/sql/virtual_schema.go
+++ b/sql/virtual_schema.go
@@ -94,9 +94,13 @@ type virtualTableEntry struct {
 	desc     *sqlbase.TableDescriptor
 }
 
-// getValuesNode returns a new valuesNode for the virtual table using the
-// provided planner.
-func (e virtualTableEntry) getValuesNode(p *planner) (*valuesNode, error) {
+type valuesNodeConstructor func(p *planner) (*valuesNode, error)
+
+// getPlanInfo returns the column metadata and a constructor for a new
+// valuesNode for the virtual table. We use deferred construction here
+// so as to avoid populating a RowContainer during query preparation,
+// where we can't guarantee it will be Close()d in case of error.
+func (e virtualTableEntry) getPlanInfo() (ResultColumns, valuesNodeConstructor) {
 	var columns ResultColumns
 	for _, col := range e.desc.Columns {
 		columns = append(columns, ResultColumn{
@@ -104,26 +108,31 @@ func (e virtualTableEntry) getValuesNode(p *planner) (*valuesNode, error) {
 			Typ:  col.Type.ToDatumType(),
 		})
 	}
-	v := p.newContainerValuesNode(columns, 0)
 
-	err := e.tableDef.populate(p, func(datums ...parser.Datum) error {
-		if r, c := len(datums), len(v.columns); r != c {
-			panic(fmt.Sprintf("datum row count and column count differ: %d vs %d", r, c))
-		}
-		for i, col := range v.columns {
-			datum := datums[i]
-			if !(datum == parser.DNull || datum.TypeEqual(col.Typ)) {
-				panic(fmt.Sprintf("datum column %q expected to be type %s; found type %s",
-					col.Name, col.Typ.Type(), datum.Type()))
+	constructor := func(p *planner) (*valuesNode, error) {
+		v := p.newContainerValuesNode(columns, 0)
+
+		err := e.tableDef.populate(p, func(datums ...parser.Datum) error {
+			if r, c := len(datums), len(v.columns); r != c {
+				panic(fmt.Sprintf("datum row count and column count differ: %d vs %d", r, c))
 			}
+			for i, col := range v.columns {
+				datum := datums[i]
+				if !(datum == parser.DNull || datum.TypeEqual(col.Typ)) {
+					panic(fmt.Sprintf("datum column %q expected to be type %s; found type %s",
+						col.Name, col.Typ.Type(), datum.Type()))
+				}
+			}
+			return v.rows.AddRow(datums)
+		})
+		if err != nil {
+			v.Close()
+			return nil, err
 		}
-		return v.rows.AddRow(datums)
-	})
-	if err != nil {
-		v.Close()
-		return nil, err
+		return v, nil
 	}
-	return v, nil
+
+	return columns, constructor
 }
 
 func (vs *virtualSchemaHolder) init(p *planner) error {

